pilot/pkg/xds/cds.go

```diff
@@ -69,20 +70,14 @@ func cdsNeedsPush(req *model.PushRequest, proxy *model.Proxy) bool {
 }
 
 func (c CdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource,
-	updates *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
+	updates *model.PushRequest) (model.Resources, model.DeletedResources, bool, model.XdsLogDetails, error) {
 	if !cdsNeedsPush(updates, proxy) {
-		return nil, model.DefaultXdsLogDetails, nil
+		return nil, nil, false, model.DefaultXdsLogDetails, nil
 	}
-	clusters, logs := c.Server.ConfigGenerator.BuildClusters(proxy, updates)
-	return clusters, logs, nil
-}
-
-// GenerateDeltas for CDS currently only builds deltas when services change. todo implement changes for DestinationRule, etc
-func (c CdsGenerator) GenerateDeltas(proxy *model.Proxy, push *model.PushContext, updates *model.PushRequest,
-	w *model.WatchedResource) (model.Resources, []string, model.XdsLogDetails, bool, error) {
-	if !cdsNeedsPush(updates, proxy) {
-		return nil, nil, model.DefaultXdsLogDetails, false, nil
+	if features.DeltaXds {
+		updatedClusters, removedClusters, logs := c.Server.ConfigGenerator.BuildDeltaClusters(proxy, updates, w)
+		return updatedClusters, removedClusters, true, logs, nil
```

Comment:
I don't think `true` is always correct here. BuildDeltaClusters can produce deltas OR sotw. if it does produce sotw, then we need the ads logic to automatically determine the removedClusters.

Reply:
Also this brings up an interesting point - I am pretty sure this means we will be missing removals. But no tests will ever fail when we miss removals...

Reply:
comment still applies

Reply:
even after moving the logic to cds? Are you saying in future it can still return sotw and hence we need to have that flag of usedDelta?